File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/KeyDerivation.kt
package com.blockchaincommons.bccomponents

/**
 * Interface for key derivation implementations.
 *
 * Each implementation derives a symmetric key from a secret and uses that
 * derived key to lock (encrypt) or unlock (decrypt) a content key. The
 * derivation parameters are serialised as CBOR and stored in the encrypted
 * message's Additional Authenticated Data (AAD) so that unlock can recover
 * them.
 */
interface KeyDerivation {
    /**
     * Derives a key from [secret], then encrypts [contentKey] with it.
     *
     * The CBOR-encoded derivation parameters are placed in the AAD of the
     * returned [EncryptedMessage].
     */
    fun lock(contentKey: SymmetricKey, secret: ByteArray): EncryptedMessage

    /**
     * Derives a key from [secret], then decrypts [encryptedMessage] to
     * recover the original content key.
     */
    fun unlock(encryptedMessage: EncryptedMessage, secret: ByteArray): SymmetricKey
}

File: typescript/bc-components/src/encrypted-key/key-derivation.ts
import type { Cbor } from '@bc/dcbor';

import type { EncryptedMessage } from '../symmetric/encrypted-message.js';
import type { SymmetricKey } from '../symmetric/symmetric-key.js';

export interface KeyDerivation {
    index(): number;
    lock(contentKey: SymmetricKey, secret: Uint8Array): EncryptedMessage;
    unlock(encryptedMessage: EncryptedMessage, secret: Uint8Array): SymmetricKey;
    toCbor(): Cbor;
}
